### ISSUE TITLE
Retry policy in service-router is not working as expected

### ISSUE THREAD

**Issue by arushi315** [hashicorp/consul#10274]
Hi Team, 
  
I am using ingress gateway for my service in consul service mesh with ring_hash load balancing policy and I am facing issue with the retry policy. Here is how my configuration for service-resolver and service-router looks like: 

``` json
{
            "Kind": "service-resolver",
            "Name": "myservice",
            "LoadBalancer": {
                "Policy": "ring_hash",
                "HashPolicies": [
                    {
                        "Field": "query_parameter",
                        "FieldValue": "myserviceid",
                        "Terminal": true
                    },
                    {
                        "Field": "header",
                        "FieldValue": "myserviceid", 
                        "Terminal": true
                    }
                ]
            }
        } 
```

``` json
{
    "Kind": "service-router",
    "Name": "ingress-http-backend",
    "Routes": [
        {
            "Match": {
                "HTTP": {
                    "PathPrefix": "/myservice"
                }
            },
            "Destination": {
                "RequestTimeout": "30m0s",
                "Service": "myservice",
                "NumRetries": 3,
                "RetryOnConnectFailure": true,
                "RetryOnStatusCodes": [
                    503,
                    504
                ]
            }
        }
    ],
    "CreateIndex": 160,
    "ModifyIndex": 251
} 
```

snippet of envoy config: 
```
"dynamic_route_configs": [
    {
     "version_info": "00000009",
     "route_config": {
      "@type": "type.googleapis.com/envoy.config.route.v3.RouteConfiguration",
      "name": "8080",
      "virtual_hosts": [
       {
        "name": "ingress-http-backend",
        "domains": [
         "*",
         "*:8080"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/myservice"
          },
          "route": {
           "cluster": "myservice.default.dev-setup.internal.1c0cc154-47ba-2984-0f53-4dc905625cff.consul",
           "timeout": "1800s",
           "retry_policy": {
            "retry_on": "connect-failure,retriable-status-codes",
            "num_retries": 3,
            "retriable_status_codes": [
             503,
             504
            ]
           },
           "hash_policy": [
            {
             "terminal": true,
             "query_parameter": {
              "name": "myserviceid"
             }
            },
            {
             "header": {
              "header_name": "myserviceid"
             },
             "terminal": true
            }
           ]
          }
         },
         {
          "match": {
           "prefix": "/"
          },
          "route": {
           "cluster": "ingress-http-backend.default.dev-setup.internal.1c0cc154-47ba-2984-0f53-4dc905625cff.consul"
          }
         }
        ]
       }
      ]
```
I have deployed 2 nodes for myservice and I am making requests to both nodes with request header 'myserviceid' for ring_hash policy. In order to test the retry logic, I bring down one node while the request is already connected. 

Expecting: I am expecting that the request will be retried with the other node that is still up and running 

Actual: I instead get a 503 response with message "upstream connect error or disconnect/reset before headers. reset reason: connection termination"

Snippet of ingress logs: 
```
[2021-05-18 16:23:30.401][13][debug][http] [source/common/http/filter_manager.cc:839] [C108][S17841730658672010268] Sending local reply with details upstream_reset_before_response_started{connection termination}
[2021-05-18 16:23:30.401][13][debug][http] [source/common/http/conn_manager_impl.cc:1484] [C108][S17841730658672010268] encoding headers via codec (end_stream=false):
':status', '503'
'content-length', '95'
'content-type', 'text/plain'
'date', 'Tue, 18 May 2021 16:23:30 GMT'
'server', 'envoy'
```

Versions: 
Consul - 1.10.0 beta1 
Envoy - 1.17.0
Nomad - 1.0.4

Is there any configuration that I am may be missing or is this an issue? Would really appreciate any guidance on this. 
Please let me know if anything else is needed for troubleshooting. 

Thank you!

**Comment by idrennanvmware** [comment]
@arushi315  - just to add a little more to your issue.  As I understand it you have the following scenario:

Client App -> Ingress -> Your service

The client app does long polls (2+ mins) which you have solved by upgrading to the Beta Consul, in addition you also have a routing "sticky" requirement to your container (hence the ring_hash).  Now you are testing a scenario where a client is connected to a container (works right now) but the container is killed mid poll - correct?

So I have a question about the client app - does it ever resolve correctly?  Or is it always stuck in this state?  Does the client app have retry logic, or is the connection immediately reattempted and fails?

**Comment by arushi315** [comment]
Hi @idrennanvmware, 

Actual long polling clients should ideally reconnect and the clients that are sending the notification or query may or may not have a retry logic. 
For my testing I am simulating my requests from POSTMAN. 

With the retry in ingress, I am looking to reduce the failures along with other measures (yet to figure out).

**Comment by blake** [comment]
I believe the 503 is being returned by the local proxy, and thus would not be be subject to the retry policy which is evaluating actual responses from the upstream.

Based on the error message  "upstream connect error or disconnect/reset before headers. reset reason: connection termination", I suspect the connection would be successfully retried if `retry_policy.retry_on` included the `reset` condition.

> **reset**
> Envoy will attempt a retry if the upstream server does not respond at all (disconnect/reset/read timeout.)

Unfortunately Consul does not currently allow enabling this retry condition, although you should able to instruct Envoy to use it by sending the [x-envoy-retry-on](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#x-envoy-retry-on) header in the request from your application. For example:

```shell
curl --header "x-envoy-retry-on: reset" http://<ip_or_hostname>/myservice
```

Could you try this out and let us know if it resolves the issue? If it does, we can open a separate issue to track exposing `retry` as a condition in the service router config entry.

**Comment by arushi315** [comment]
Thanks @Blake for looking into this. 
Let me give [x-envoy-retry-on](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#x-envoy-retry-on) a try and I will get back to you.

**Comment by arushi315** [comment]
Hi @Blake,

I want to clarify on one thing - when I say I shutdown one node, I am actually reducing the nomad job count to 1 and in that case I believe local proxy will shutdown as well and not just the service container. 

And I tried `x-envoy-retry-on` request header with the following values: 
`reset`
`5xx` (which should cover gateway-error and connect-failure)

No luck with retry with these.

**Comment by david-yu** [comment]
Hi @arushi315 it looks like this was addressed via https://github.com/hashicorp/consul/pull/12385, could you upgrade to 1.11.4 or 1.10.9 see if this still an issue?

**Comment by arushi315** [comment]
Hi @david-yu 
I tested the retry logic with Consul 1.10.9 and here is what I tried: 
Instances: 
1 ingress 
2 my service

Using service-router configuration: 
```
                "NumRetries": 3,
                "RetryOnConnectFailure": true,
                "RetryOnStatusCodes": [
                    500,501,502,503,504
                ]
``` 
I made 2 long running requests to my service, each request connected to separate instance. 
I exec into my container and killed the service on on the instance. 
The request that was connected to the killed service, I received 503. 

I increased the number of retries to 5 and performed the same steps. This time the retry actually and the request was retried with the instance that up all the time. 
I am little confuse here as if it was going to retry with the other node anyway why did I have to increase the number of retries. 

Retry ingress log snippet: 
```
[2022-04-08 21:22:21.099][1][debug][upstream] [source/common/upstream/upstream_impl.cc:256] transport socket match, socket default selected for host with address 172.26.64.1:29999
[2022-04-08 21:22:21.099][1][debug][upstream] [source/common/upstream/strict_dns_cluster.cc:177] DNS refresh rate reset for 172.26.64.1, refresh rate 5000 ms
[2022-04-08 21:22:24.513][15][debug][router] [source/common/router/router.cc:1285] [C241][S1660965084484957313] upstream headers complete: end_stream=false
[2022-04-08 21:22:24.514][15][debug][router] [source/common/router/upstream_request.cc:334] [C241][S1660965084484957313] resetting pool request
[2022-04-08 21:22:24.538][15][debug][router] [source/common/router/router.cc:1670] [C241][S1660965084484957313] performing retry
[2022-04-08 21:22:24.542][15][debug][router] [source/common/router/upstream_request.cc:416] [C241][S1660965084484957313] pool ready
[2022-04-08 21:22:24.544][15][debug][router] [source/common/router/router.cc:1285] [C241][S1660965084484957313] upstream headers complete: end_stream=false
[2022-04-08 21:22:24.544][15][debug][router] [source/common/router/upstream_request.cc:334] [C241][S1660965084484957313] resetting pool request
[2022-04-08 21:22:24.588][15][debug][router] [source/common/router/router.cc:1670] [C241][S1660965084484957313] performing retry
[2022-04-08 21:22:24.590][15][debug][router] [source/common/router/upstream_request.cc:416] [C241][S1660965084484957313] pool ready
[2022-04-08 21:22:24.591][15][debug][router] [source/common/router/router.cc:1285] [C241][S1660965084484957313] upstream headers complete: end_stream=false
[2022-04-08 21:22:24.591][15][debug][router] [source/common/router/upstream_request.cc:334] [C241][S1660965084484957313] resetting pool request
[2022-04-08 21:22:24.606][15][debug][router] [source/common/router/router.cc:1670] [C241][S1660965084484957313] performing retry
[2022-04-08 21:22:24.607][15][debug][router] [source/common/router/upstream_request.cc:416] [C241][S1660965084484957313] pool ready
[2022-04-08 21:22:24.608][15][debug][router] [source/common/router/router.cc:1285] [C241][S1660965084484957313] upstream headers complete: end_stream=false
[2022-04-08 21:22:24.608][15][debug][router] [source/common/router/upstream_request.cc:334] [C241][S1660965084484957313] resetting pool request
[2022-04-08 21:22:24.729][15][debug][router] [source/common/router/router.cc:1670] [C241][S1660965084484957313] performing retry
[2022-04-08 21:22:24.730][15][debug][router] [source/common/router/upstream_request.cc:416] [C241][S1660965084484957313] pool ready
[2022-04-08 21:22:24.731][15][debug][router] [source/common/router/router.cc:1285] [C241][S1660965084484957313] upstream headers complete: end_stream=false
[2022-04-08 21:22:24.731][15][debug][router] [source/common/router/upstream_request.cc:334] [C241][S1660965084484957313] resetting pool request
[2022-04-08 21:22:24.732][1][debug][upstream] [source/common/upstream/cluster_manager_impl.cc:1087] membership update for TLS cluster myservice.default.dev-setup.internal.d2a007e5-7317-3311-1865-1baacfb82061.consul added 0 removed 0
[2022-04-08 21:22:24.732][1][debug][upstream] [source/common/upstream/cluster_manager_impl.cc:1093] re-creating local LB for TLS cluster myservice.default.dev-setup.internal.d2a007e5-7317-3311-1865-1baacfb82061.consul
[2022-04-08 21:22:24.732][15][debug][upstream] [source/common/upstream/cluster_manager_impl.cc:1087] membership update for TLS cluster myservice.default.dev-setup.internal.d2a007e5-7317-3311-1865-1baacfb82061.consul added 0 removed 0
[2022-04-08 21:22:24.732][1][debug][upstream] [source/common/upstream/cluster_manager_impl.cc:788] host 0xf9fff70bd80 in cluster myservice.default.dev-setup.internal.d2a007e5-7317-3311-1865-1baacfb82061.consul was ejected by the outlier detector
[2022-04-08 21:22:24.732][15][debug][upstream] [source/common/upstream/cluster_manager_impl.cc:1093] re-creating local LB for TLS cluster myservice.default.dev-setup.internal.d2a007e5-7317-3311-1865-1baacfb82061.consul
[2022-04-08 21:22:24.920][15][debug][router] [source/common/router/router.cc:1670] [C241][S1660965084484957313] performing retry
[2022-04-08 21:22:24.923][15][debug][router] [source/common/router/upstream_request.cc:416] [C241][S1660965084484957313] pool ready
[2022-04-08 21:22:25.819][15][debug][http] [source/common/http/conn_manager_impl.cc:274] [C361] new stream
[2022-04-08 21:22:25.819][15][debug][http] [source/common/http/conn_manager_impl.cc:867] [C361][S15029791713036707267] request headers complete (end_stream=true):
```

**Comment by blake** [comment]
Support for configuring `retry_on` was added by PR #12890. This feature will be available in Consul 1.14.
